### PR TITLE
[master] Add option to dump contract code for data_migrate

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

- Enabling code dumping for all contracts in persistence in offline.

## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [ ] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libData/AccountData/AccountStore.cpp [hunk 1]
```diff
@@ -608,5 +608,31 @@ bool AccountStore::MigrateContractStates2(
...
+    Account account;
+    if (!account.DeserializeBase(bytes(i.second.begin(), i.second.end()), 0)) {
+      LOG_GENERAL(WARNING, "Account::DeserializeBase failed");
+      return false;
+    }
+
+    if (account.isContract()) {
+      account.SetAddress(address);
+      std::ofstream os("contracts/" + account.GetAddress().hex() +
+                       CONTRACT_FILE_EXTENSION);
```
Comment (jjcnn) [inline src/libData/AccountData/AccountStore.cpp:629]: We'll very soon be able to release the external libraries feature, so this should be something like

```
(is_library ? LIBRARY_CODE_EXTENSION : CONTRACT_FILE_EXTENSION)
```
I can't remember how to determine `is_library` from an Account, but I've seen logic that determines it somewhere else, so it must be possible.

**Review by jjcnn — CHANGES_REQUESTED** [review]
I'm getting

```
$ ~/Projects/Zilliqa/build/bin/data_migrate -b -a
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::property_tree::ptree_bad_path> >'
  what():  No such node (node.general.NUM_EPOCHS_PER_PERSISTENT_DB)
Aborted (core dumped)
```
when I try to run this. It worked before - have I messed something up in the setup?

File: src/libData/AccountData/AccountStore.cpp (hunk 1 above)
Comment (kaikawaliu) [reply src/libData/AccountData/AccountStore.cpp:629]: oh okay, it's possible but may need some extra effort to do so. but seems we haven't library code yet, maybe we can live with this first?

**Comment by kaikawaliu** [comment]
> I'm getting
> 
> ```
> $ ~/Projects/Zilliqa/build/bin/data_migrate -b -a
> terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::property_tree::ptree_bad_path> >'
>   what():  No such node (node.general.NUM_EPOCHS_PER_PERSISTENT_DB)
> Aborted (core dumped)
> ```
> 
> when I try to run this. It worked before - have I messed something up in the setup?

cam you try `./data_migrrate -c`, that's for your use case.

**Comment by jjcnn** [comment]
> > I'm getting
> > ```
> > $ ~/Projects/Zilliqa/build/bin/data_migrate -b -a
> > terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::property_tree::ptree_bad_path> >'
> >   what():  No such node (node.general.NUM_EPOCHS_PER_PERSISTENT_DB)
> > Aborted (core dumped)
> > ```
> > 
> > 
> > when I try to run this. It worked before - have I messed something up in the setup?
> 
> can you try `./data_migrrate -c`, that's for your use case.

```
$ ~/Projects/Zilliqa/build/bin/data_migrate -c
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::property_tree::ptree_bad_path> >'
  what():  No such node (node.general.NUM_EPOCHS_PER_PERSISTENT_DB)
Aborted (core dumped)
```
Adding options `-a` and `-b` gives the same result.

When I ran the script the other day I used the updated verson of scilla-checker, which produces different contractinfo output. I don't suppose that has changed the persistence files enough for the script no longer to work? Do I need to use a fresh set of files?

File: src/libData/AccountData/AccountStore.cpp (hunk 1 above)
Comment (jjcnn) [reply src/libData/AccountData/AccountStore.cpp:629]: Thanks.

**Comment by kaikawaliu** [comment]
> NUM_EPOCHS_PER_PERSISTENT_DB

Oh you also need to copy the `constants.xml` to the directory where the binary was run, in which 
- `ENABLE_SC` must be `true`, and `SCILLA_ROOT` points to the absolute path of your built scilla repo such as `/jacob/home/scilla_repo/scilla`.
- `ENABLE_SCILLA_MULTI_VERSION` should be false as we have only one version so far. 
- If you deal with scilla library, `DISABLE_SCILLA_LIB` must be `false`.

**Comment by jjcnn** [comment]
> > NUM_EPOCHS_PER_PERSISTENT_DB
> 
> Oh you also need to copy the `constants.xml` to the directory where the binary was run, in which
> 
>     * `ENABLE_SC` must be `true`.
> 
>     * `SCILLA_ROOT` points to the absolute path of your built scilla repo such as `/jacob/home/scilla_repo/scilla`.
> 
>     * `ENABLE_SCILLA_MULTI_VERSION` should be `false` as we have only one version so far.
> 
>     * If you deal with scilla library, `DISABLE_SCILLA_LIB` must be `false`.

That didn't help either:
```
$ ~/Projects/Zilliqa/build/bin/data_migrate -c
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::property_tree::ptree_bad_path> >'
  what():  No such node (node.general.NUM_EPOCHS_PER_PERSISTENT_DB)
Aborted (core dumped)
```

When I ran it originally I used the constants.xml from the Ziliqa repo without modification. I had to create a directory `/0` (the Scilla version) containing softlinks to the Scilla `bin` and `src` directories, but that worked fine.

**Pull request closed without merging**